src/Lean/Elab/Coinductive.lean

```diff
@@ -0,0 +1,217 @@
+  let nm := view.declName.replacePrefix topView.declName (topView.declName ++ `Shape)
+
+  let type? ← view.type?.mapM fun type => do
+    let ⟨args, out⟩ := typeToArgArr type
+    let ⟨arr, _⟩ := appsToArgArr out
+
+    let ⟨pre, post⟩ := (split topView.binders.size arr.toList).map (·.toArray) (·.toArray)
+
+    let out ← `($isTy $pre* $names* $post*)
+
+    args.reverse.foldlM (fun acc curr => `($curr → $acc)) out
+
+  return {
+    ref       := .missing
+    modifiers := view.modifiers
+    declName  := nm
+    binders   := .node .none `null (view.binders.map (·.raw))
+    type?     := type?
+  }
+
+def generateIs (vss : Array (CoInductiveView × Array Ident)) (rNameEntries : Array (Ident × Term)) : CommandElabM Unit := do
+  -- It could be worth making this extract only the names that are required.
+  let boundRNames ← rNameEntries.mapM (fun ⟨i, v⟩ => do `(bb| ( $i : $v) ))
+  let coRecArgs   ← vss.mapM (fun ⟨v, _⟩ => do `(bb| ( $(mkIdent $ v.shortDeclName) : $(←v.toRelType))))
+  let names      := vss.map  (mkIdent ·.fst.shortDeclName)
+
+  for ⟨idx, topView, argArr⟩ in vss.toList.enum.toArray do
+    let shortDeclName := topView.shortDeclName ++ `Shape
+
+    let view := {
+      ref             := .missing
+      declId          := ←`(declId| $(mkIdent shortDeclName))
+      modifiers       := topView.modifiers
+      shortDeclName
+      declName        := topView.declName ++ `Shape
+      levelNames      := topView.levelNames
+      binders         := .node .none `null $ topView.binders.append coRecArgs
+      type?           := some topView.type
+      ctors           := ←topView.ctors.mapM $ handleCtor names topView $ mkIdent shortDeclName
+      derivingClasses := #[]
+      computedFields  := #[]
+    }
+
+    trace[Elab.CoInductive] s!"{repr topView.binders}"
+    trace[Elab.CoInductive] s!"{topView.toBinderIds}"
+
+    let boundNames := rNameEntries.map Prod.fst
+    let i := boundNames[idx]! -- OK since these come from the same source
+
+    let stx ← `(command|
+      abbrev $(mkIdent $ topView.shortDeclName ++ `Is) $(topView.binders)* $boundRNames* : Prop :=
+        ∀ { $argArr* }, $i $(topView.toBinderIds)* $argArr* → $(mkIdent shortDeclName) $(topView.toBinderIds)* $boundNames* $argArr*)
+
+    trace[Elab.CoInductive] "Generating Is check:"
+    trace[Elab.CoInductive] stx
+
+    elabInductiveViews #[view]
+    elabCommand stx
+
+def elabCoInductiveViews (views : Array CoInductiveView) : CommandElabM Unit := do
+  let viewCheck ← views.mapM fun view => do
+    let ⟨tyArr, out⟩ := typeToArgArr view.type
+    let argArr := (← tyArr.mapM (fun _ => mkFreshBinderName)).map mkIdent
+
+    -- In theory we could make this handle types by simply changing the existential quantification but this would yield some pretty funny results
+    let .node _ ``Parser.Term.prop _ := out.raw | throwErrorAt out "Expected return type to be a Prop"
+    return Prod.mk view argArr
```

Comment:
How would you change the existential? The reason this works is because `Prop` is impredicative (which is a funny way of saying that `Prop -> Prop` is of type `Prop`). Once you start looking at types, you run into the issue that `Type -> Type` lives in `Type 1`, causing nasty universe bumps.

If you meant "pull in QpfTypes", sure, but that is very explicitly not something that core wants
```suggestion
    let .node _ ``Parser.Term.prop _ := out.raw | throwErrorAt out "Expected return type to be a Prop"
    return Prod.mk view argArr
```